Add doc on Token Introspection

khorne3 opened the pull request:
<!---
Notes:
- Your PR should conform to our [Contributing Guidelines](https://github.com/auth0/docs/blob/master/CONTRIBUTING.md)
- If applicable, add details to the [update feed](https://github.com/auth0/docs/tree/master/updates)
- Make sure your PR gets deployed to a Heroku [Review App](https://github.com/auth0/docs/blob/master/CONTRIBUTING.md#review-apps) without errors.
- Please include a short description
- If your PR is a work in progress title it [DO NOT MERGE]
--->

Commit:
Add intro and sample HAR snippet

Commit:
Add info on providing public key to Auth0

Commit:
Add info on calling Auth API

jhiner commented on articles/api-auth/token-introspection.md:
```diff
@@ -0,0 +1,81 @@
+
+# Token Introspection
+
+Auth0 supports [OAuth 2.0 Token Introspection](https://tools.ietf.org/html/rfc7662), which allows you to ask Auth0 for details about any access tokens presented to your API. In addition to asking Auth0 about the *state* of the access token (whether it is active/expired/revoked), you can request information about the token's scopes and the context in which it was issued (such as who issued the tok ...
```
In addition to asking Auth0 about the *state* of the access token (whether it is active or not)

jhiner commented on articles/api-auth/token-introspection.md:
```diff
@@ -0,0 +1,81 @@
+
+1. You can define the `verificationKey` on the API's Auth0 record. This is a JSON Web Key (JWK) or PEM-encoded certificate.
+2. You can define the `verificationLocation` on the API's Auth0 record. This is a URI from which Auth0 fetches JSON Web Keys (JWK).
+3. If the API's `identifier` is `http:`- or `https:`-based (such as `audience=https://example.com/foo`), then Auth0 attempts to fetch the key from the well-known JWK registry of the host (for the provided URI example, Auth0 attempts to fetch from `https://example.com/.well-known/jwks.json`).
```
in this case it wouldn't be audience=https:, it's referring to the identifier. 

So simply (such as `https://example.com/foo`)

jhiner commented on articles/api-auth/token-introspection.md:
````diff
@@ -0,0 +1,81 @@
+* **Client Assertion Type**: Set this to `client_assertion_type: urn:ietf:params:oauth:client-assertion-type:jwt-bearer`;
+* **Client Assertion**: a JWT signed with your API's private key so that Auth0 can validate its signature using the associated public key.
+
+```har
````
For some reason this isn't showing up correctly - I don't see the `postData`. 

Also, maybe it's better to add `token` to the list of params above to briefly explain that it is the literal token value.

jhiner commented:
Is it possible to add a link to the sidebar under API-auth, Dynamic Client Registration:

https://auth0-docs-content-pr-3985.herokuapp.com/docs/api-auth

Commit:
Incorporate feedback

Commit:
Test HAR snippet

Commit:
Remove HAR

Commit:
Add link to sidebar

Commit:
Fix curl snippet

Commit:
Fix line break

khorne3 commented:
@jhiner This is ready for (re)review. I added a link to the sidebar, and I changed the second curl snippet (turns out our HAR generator only works with a JSON body right now).

jhiner commented on articles/api-auth/token-introspection.md:
```diff
@@ -0,0 +1,70 @@
+
+# Token Introspection
+
+Auth0 supports [OAuth 2.0 Token Introspection](https://tools.ietf.org/html/rfc7662), which allows you to ask Auth0 for details about any access tokens presented to your API. In addition to asking Auth0 about the *state* of the access token (whether it is active or not), you can request information about the token's scopes and the context in which it was issued (such as who issued the token and to ...
```
One more minor change, like this:
> Auth0 supports [OAuth 2.0 Token Introspection](https://tools.ietf.org/html/rfc7662), which allows a protected resource to query an Authorization Server (Auth0) to find out about the status of an OAuth 2.0 token (whether it is active or not). You can also obtain information such as the token's scopes and the context in which it was issued (such as who issued the token and to whom the token was issued).

Commit:
Fix intro

jhiner commented on articles/api-auth/token-introspection.md:
````diff
@@ -0,0 +1,69 @@
+  --data 'token=TOKEN_VALUE&client_assertion_type=ASSERTION_TYPE&client_assertion=ASSERTION'
+```
+
+You can use token introspection to request information for **access** and **refresh** tokens.
````
You can use token introspection to request information for JWT and non-JWT access tokens.

mpaktiti commented on articles/api-auth/token-introspection.md:
```diff
@@ -0,0 +1,69 @@
+
+# Token Introspection
+
+Auth0 supports [OAuth 2.0 Token Introspection](https://tools.ietf.org/html/rfc7662), which allows a protected resource to query an Authorization Server (Auth0) to request the status of an OAuth 2.0 token (whether it is active or not). The protected resource can also obtain information such as the token's scopes and the context in which it was issued (such as who issued the token and to whom the t ...
```
If you don't know what token introspection is (I didn't), this intro leaves you with questions.
- In which case would I use this?
- Does using it means that my API doesn't have to [verify the token](https://auth0.com/docs/api-auth/tutorials/verify-access-token)?
- In the case of Auth0, where the access_token is a JWT, what is our recommendation to our customers? Use this or validate the tokens themselves?

In addition to my questions, I would expect something like this in the intro https://tools.ietf.org/html/rfc7662#section-1

mpaktiti commented on articles/api-auth/token-introspection.md:
```diff
@@ -0,0 +1,69 @@
+2. You can define the `verificationLocation` on the API's Auth0 record. This is a URI from which Auth0 fetches JSON Web Keys (JWK).
+3. If the API's `identifier` is `http:`- or `https:`-based (such as `https://example.com/foo`), then Auth0 attempts to fetch the key from the well-known JWK registry of the host (for the provided URI example, Auth0 attempts to fetch from `https://example.com/.well-known/jwks.json`).
+
+If you choose to provide `verificationKey` or `verificationLocation` values, you can provide update your API's Auth0 record by making the appropriate `POST` call to the [Management API](/api/management/v2#!/Resource_Servers/patch_resource_servers_by_id).
```
The Management API docs do not list these two attributes: `verificationKey` and `verificationLocation`. Can you update it? Or ask Angela to do it.

mpaktiti commented on articles/api-auth/token-introspection.md:
```diff
@@ -0,0 +1,69 @@
+	"cookies": [],
+	"headers": [{
+		"name": "Authorization",
+		"value": "Bearer ABCD"
```
Replace `ABCD` with sth like `MGMT_API_ACCESS_TOKEN` and add a note after the har that links to the doc on how to get one.

mpaktiti commented on articles/api-auth/token-introspection.md:
```diff
@@ -0,0 +1,69 @@
+
+## Calling the Authorization API's Token Introspection Endpoint
+
+To use token introspection, make a `POST` call to the [Authorization API](/api/authentication) where the content type is `application/x-www-form-urlencoded` and you've included the following parameter value pairs as part of the payload:
```
Replace `Authorization API` with `Authentication API`. Also, shouldn't this PR add the `/oauth/introspect` endpoint to the API doc?

mpaktiti commented on articles/api-auth/token-introspection.md:
```diff
@@ -0,0 +1,69 @@
+
+* **Token**: The literal value of the token;
+* **Client Assertion Type**: Set this to `client_assertion_type: urn:ietf:params:oauth:client-assertion-type:jwt-bearer`;
+* **Client Assertion**: a JWT signed with your API's private key so that Auth0 can validate its signature using the associated public key.
```
I'm not clear on what this `Client Assertion` is. Elaborate maybe?

mpaktiti commented on articles/api-auth/token-introspection.md:
````diff
@@ -0,0 +1,69 @@
+```curl
+curl --request POST \
+  --url 'https://${account.namespace}/oauth/introspect' \
+  --header 'authorization: Bearer ABCD' \
````
Replace `ABCD` with sth like `MGMT_API_ACCESS_TOKEN` and add a note after the har that links to the doc on how to get one.

mpaktiti commented on articles/api-auth/token-introspection.md:
````diff
@@ -0,0 +1,69 @@
+  --data 'token=TOKEN_VALUE&client_assertion_type=ASSERTION_TYPE&client_assertion=ASSERTION'
+```
+
+You can use token introspection to request information for **JWT access** tokens, **non-JWT access** tokens, and **refresh** tokens.
````
Link to the docs maybe?
- JWT access token -> https://auth0.com/docs/tokens/access-token
- Refresh token -> https://auth0.com/docs/tokens/preview/refresh-token

mpaktiti commented on config/sidebar.yml:
```diff
@@ -49,6 +49,9 @@ articles:
 
+      - title: "Token Introspection"
+        url: "/api-auth/token-introspection"
+
```
Can you also add an entry for this doc at the api-auth index? https://auth0.com/docs/api-auth
Same place, after `Dynamic Client Registration`.

mpaktiti reviewed (changes requested):
Hey I left some comments.

Commit:
Add details

Commit:
Add links

Commit:
Add update

Commit:
Add endpoint to Auth API

Commit:
Add status indicators

jhiner commented:
@khorne3 
To answer your questions from Slack:
> Can you please provide some use cases?
Token introspection is used by Resource Servers to verify the validity of tokens. Note that we need to remove refresh tokens from the end of the doc. We technically support it, but we don't have a way for OAuth clients to call introspection right now. So for now, it's just JWT and non-JWT access tokens. 

The main use case we built introspection for was App Specific Passwords (ASP). In this case, the ASP is a non-JWT token and the API would use token introspection to verify the validity of the token. I am preparing some information to send you about the ASP feature. Perhaps we should roll out token introspection docs alongside ASP (another section along with Dynamic Client Reg and Token Introspection in the sidebar).

In case of access tokens, because we only issue JWTs today, introspection isn't super useful for that. In the future, we will start issuing non-JWT reference tokens (much like the opaque `userinfo` tokens today), and in that case, introspection is the mechanism by which APIs would verify their validity.

> Should the API still verify the token, or are there cases where this isn’t necessary because token introspection provides sufficient info?

Token introspection is a means by which an API can verify a token. 

> If the access token is a JWT, what’s our best practice recommendation: use token introspection or validate the tokens themselves?

It's fine to continue to use existing token validation mechanisms with JWT tokens. As mentioned above, this is more useful for newer features we will roll out in the future (e.g. non-JWT reference tokens).

Commit:
Add intro

khorne3 commented:
See https://github.com/auth0/docs/pull/4053 for new PR that will include ASP docs.

chenkie commented on articles/api-auth/token-introspection.md:
```diff
@@ -0,0 +1,83 @@
+
+Auth0 supports [OAuth 2.0 Token Introspection](https://tools.ietf.org/html/rfc7662), which allows a protected resource to query an Authorization Server (Auth0) to request the status of an OAuth 2.0 token (whether it is active or not). The protected resource can also obtain information such as the token's scopes and the context in which it was issued (such as who issued the token and to whom the t ...
+
+Because token instrospection is a means by which an API can verify a token, you do not need to configure any addition verification flows performed by the API. Additionally, for any existing authorization flows involving JWT access tokens, you do not need to update your processes to use token introspection.
```
typo "addition"
